model/riscv_vmem.sail

```diff
@@ -0,0 +1,855 @@
+  pte_PPNs_size_bits  : {22, 44},        //   22    44    44
+  pte_PPN_j_size_bits : {10,  9}         //   10     9     9
+}
+
+// PRIVATE
+let sv32_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 9,
+       satp_asid_lsb_index = 22,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 22,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 32,
+       vpn_size_bits       = 10,
+
+       // PTE
+       levels              = 2,
+       log_pte_size_bytes  = 2,    // 4 Bytes
+       pte_msbs_lsb_index  = 32,
+       pte_msbs_size_bits  = 0,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 22,
+       pte_PPN_j_size_bits = 10
+}
+
+// PRIVATE
+let sv39_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 39,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 3,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// PRIVATE
+let sv48_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 48,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 4,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// For future; not currently used
+// PRIVATE
+/*
+let sv57_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits        = 57,
+       vpn_size_bits       = 9,
+
+       // PTE
+       levels              = 5,
+       log_pte_size_bytes  = 3,    // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+*/
+
+// PRIVATE
+val      undefined_SV_Params : unit -> SV_Params
+
+function undefined_SV_Params() = sv32_params
+
+// ****************************************************************
+// Fields of virtual addresses
+
+// PRIVATE: Extract full VPN field from VA
+function vpns_of_va(sv_params : SV_Params,
+                    va        : bits(64)) -> bits(64) = {
+    let mask : bits(64) = (zero_extend(0b1) << sv_params.va_size_bits) - zero_extend(0b1);
```

Comment:
Ok, thanks for the suggestion.

Reply:
I’ve incorporated this successfully for all the (many) mask computations.

But there are two places where typechecking fails, both involving PAGESIZE_BITS : nat.
One place is the following: Both versions 2 and 3 fail on typecheck ("Could not resolve quantifiers for zero_extend”)

    function offset_of_va(va : bits(64)) -> bits(64) = {
        let mask : bits(64) = (zero_extend(0b1) << PAGESIZE_BITS) - 1;    // 1: original code; typecheck succeeds
        // let mask : bits(64) = zero_extend(ones(PAGESIZE_BITS));           // 2: typecheck fails
        // let mask : bits(64) = zero_extend(64, ones(PAGESIZE_BITS));    // 3: typecheck fails
        va & mask
    }

Reply:
Presumably if you constrain PAGESIZE_BITS's type to not be a nat then it'll be happy, because it's (rightly) upset that you can't zero_extend e.g. a 12356-bit number to 64 bits.

Reply:
Ah, I see (even though PAGESIZE_BIT is an immutable constant with value 12?).
I changed the declaration of PAGESIZE_BITS from:

    let PAGESIZE_BITS : nat = 12

to this:

    let PAGESIZE_BITS : {12} = 12

and it fixed the problem.
Thanks!

Reply:
A pattern we use in sail-cheri-riscv for this kind of thing is:

```
type PAGESIZE_BITS : Int = 12
let  PAGESIZE_BITS = sizeof(PAGESIZE_BITS)
```

which avoids duplicating the value but instead is rather more longwinded. Perhaps there is (or could be invented) some nice syntax for this kind of case.